src/main/java/com/linkedin/kmf/services/MultiClusterTopicManagementService.java

```diff
@@ -0,0 +1,343 @@
+        ? (Map) props.get(MultiClusterTopicManagementServiceConfig.CONFIG_PER_CLUSTER_CONFIG) : new HashMap<>();
+
+    for (Map<String, Object> serviceProp: _serviceProps.values()) {
+      if (serviceProp.containsKey(MultiClusterTopicManagementServiceConfig.TOPIC_CONFIG))
+        throw new ConfigException("The raw per-cluster config for MultiClusterTopicManagementService must not contain " +
+            MultiClusterTopicManagementServiceConfig.TOPIC_CONFIG);
+      serviceProp.put(MultiClusterTopicManagementServiceConfig.TOPIC_CONFIG, topic);
+    }
+
+    _scheduleIntervalMs = config.getInt(MultiClusterTopicManagementServiceConfig.REBALANCE_INTERVAL_MS_CONFIG);
+
+    _executor = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+      @Override
+      public Thread newThread(Runnable r) {
+        return new Thread(r, _serviceName + "-multi-cluster-topic-management-service-thread");
+      }
+    });
+  }
+
+  @Override
+  public synchronized void start() {
+    if (_isRunning.compareAndSet(false, true)) {
+      Runnable r = new TopicManagementRunnable();
+      _executor.scheduleWithFixedDelay(r, 0, _scheduleIntervalMs, TimeUnit.MILLISECONDS);
+      LOG.info("{}/MultiClusterTopicManagementService started.", _serviceName);
+    }
+  }
+
+  @Override
+  public synchronized void stop() {
+    if (_isRunning.compareAndSet(true, false)) {
+      _executor.shutdown();
+      LOG.info("{}/MultiClusterTopicManagementService stopped.", _serviceName);
+    }
+  }
+
+  @Override
+  public boolean isRunning() {
+    return _isRunning.get() && !_executor.isShutdown();
+  }
+
+  @Override
+  public void awaitShutdown() {
+    try {
+      _executor.awaitTermination(Integer.MAX_VALUE, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      LOG.info("Thread interrupted when waiting for {}/MultiClusterTopicManagementService to shutdown", _serviceName);
+    }
+    LOG.info("{}/MultiClusterTopicManagementService shutdown completed", _serviceName);
+  }
+
+  private class TopicManagementRunnable implements Runnable {
+    @Override
+    public void run() {
+      try {
+        for (Map props : _serviceProps.values()) {
+          maybeCreateTopic(props);
+        }
+
+        maybeAddPartitions(_serviceProps.values());
+
+        for (Map.Entry<String, Map> entry : _serviceProps.entrySet()) {
+          String clusterName = entry.getKey();
+          Map props = entry.getValue();
+          try {
+            maybeReassignPartitionAndElectLeader(props);
+          } catch (IOException | ZkNodeExistsException | AdminOperationException e) {
+            LOG.warn(_serviceName + "/MultiClusterTopicManagementService will retry later in cluster " + clusterName, e);
+          }
+        }
+      } catch (Exception e) {
+        LOG.error(_serviceName + "/MultiClusterTopicManagementService will stop due to error.", e);
+        stop();
+      }
+    }
+  }
+
+  private static void maybeCreateTopic(Map props) throws Exception {
```

Comment:
This props is the same props that we pass to the constructor of TopicManagementService. But I can not pass TopicManagementServiceConfig because we need to get configs such as TOPIC_FACTORY_PROPS_CONFIG directly from the Map. Alternative we can pass all parameters (e.g. topic, zkconnect) as parameters to these methods but it makes the signature of these methods very clumsy. What would you recommend?